apps/story_qa/langchain_qa.py

```diff
@@ -0,0 +1,108 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import time
+
+from langchain.callbacks.manager import CallbackManager
+from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
+from langchain.chains.qa_with_sources import load_qa_with_sources_chain
+from langchain.embeddings.huggingface import HuggingFaceEmbeddings
+from langchain.llms import GPT4All
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+from langchain.vectorstores.faiss import FAISS
+from tqdm import tqdm
+from util import download_story
+
+
+def download_model(local_path):
+    from pathlib import Path
+
+    import requests
+    from tqdm import tqdm
+
+    Path(local_path).parent.mkdir(parents=True, exist_ok=True)
+
+    # skip downloading the model
+    if Path(local_path).exists():
+        print("Skip downloading as model already exists")
+        return
+    # Example model. Check https://github.com/nomic-ai/gpt4all for the latest models.
+    url = "http://gpt4all.io/models/ggml-gpt4all-j-v1.3-groovy.bin"
+
+    # send a GET request to the URL to download the file. Stream since it's large
+    response = requests.get(url, stream=True)
+
+    # open the file in binary mode and write the contents of the response to it in chunks
+    # This is a large file, so be prepared to wait.
+    with open(local_path, "wb") as f:
+        for chunk in tqdm(response.iter_content(chunk_size=8192)):
+            if chunk:
+                f.write(chunk)
+
+
+print("Downloading model ...")
+model_path = "./ggml-gpt4all-j-v1.3-groovy.bin"
+download_model(model_path)
+
+# Download The Project Gutenberg eBook of War and Peace
+print("Downloading story ...")
+book_path = download_story()
+
+# setup the model
+print("Setting up langchain ...")
+callback_manager = CallbackManager([StreamingStdOutCallbackHandler()])
+llm = GPT4All(
+    model=model_path,
+    # backend="gptj",
+    callbacks=callback_manager,
+    verbose=True,
+    n_threads=16,
+)
+
+# setup the hugging face embeddding model
+embedding_model_name = "sentence-transformers/all-MiniLM-L6-v2"
+model_kwargs = {"device": "cuda"}
+hf_embeddings = HuggingFaceEmbeddings(
+    model_name=embedding_model_name,
+    model_kwargs=model_kwargs,
+)
+
+text_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=50)
+texts = text_splitter.split_text(open(book_path).read())
+
+# create FAISS vector index based on HuggingFace Embeddings
+print("Building FAISS Vector Index ...")
+ss = time.perf_counter()
+store = FAISS.from_texts(
+    texts,
```

Comment:
One minor thing if you want to have separate feature extraction and index creation steps with timing.

Reply:
It takes around `1.5s` to build the index in langchain, whereas we spend like `23s`. I'm going to profile what is going on here. 
Possible reasons: 
1. Low batch size
2. Sqlalchemy overhead in create table.

Reply:
We are creating a secondary table if I recall correctly. That may also lead to extra overhead.

Reply:
That is part of the `create_vector_index` call, which is actually fast.

Reply:
Create table  12919.144 ms
Extract features 28384.113 ms
Create index 326.020 ms

Reply:
@jiashenC Please verify the numbers for `langchain`

Reply:
Are these numbers from EVA with ray enabled? Could we get also numbers with ray off?

Reply:
Yes
Without ray:
Create table 13136.788 ms 
Extract features 10273.294 ms
Create index  282.686 ms